src/transformers/modeling_utils.py

```diff
@@ -1402,6 +1433,42 @@ def _get_generated_ngrams(hypo_idx):
     return banned_tokens
 
 
+def calc_banned_bad_words_tokens(prev_input_ids, bad_words_tokens):
+    banned_tokens = []
+
+    def _tokens_match(prev_tokens, tokens):
+        if len(tokens) == 0:
+            # if bad word tokens is just one token always ban it
+            return True
+        if len(tokens) > len(prev_input_ids):
+            # if bad word tokens are longer then prev input_ids they can't be equal
+            return False
+
+        if prev_tokens[-len(tokens) :] == tokens:
+            # if tokens match
+            return True
+        else:
+            return False
+
+    for prev_input_ids_slice in prev_input_ids:
+        banned_tokens_slice = []
+
+        for banned_token_seq in bad_words_tokens:
+            assert len(banned_token_seq) > 0, "Banned words token sequences {} cannot have an empty list".format(
+                bad_words_tokens
+            )
+
+            if _tokens_match(prev_input_ids_slice.tolist(), banned_token_seq[:-1]) is False:
```

Comment:
Ok seems good for now. At some point in the future, I guess we will want to make generation faster by keeping everything in tensors instead of converting to lists but probably we will need to think deeply about how to make it fast anyway at that time (e.g. switching to a low-level language).